### PR TITLE
Stax support for Exchange

### PR DESCRIPTION
**Pull request by fbeutin-ledger**

### CODE REVIEW

File: src/ui/menu_nbgl.c [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+static const char* const info_types[] = {"Version", "Exchange App"};
+static const char* const info_contents[] = {APPVERSION, "(c) 2023 Ledger"};
+
+#define SETTINGS_PAGE_NUMBER 2
+static bool nav_callback(uint8_t page, nbgl_pageContent_t* content) {
+    if (page == 0) {
+        content->type = INFOS_LIST;
+        content->infosList.nbInfos = 2;
+        content->infosList.infoTypes = (const char**) info_types;
+        content->infosList.infoContents = (const char**) info_contents;
```
Comment (xchapron-ledger) [inline src/ui/menu_nbgl.c:27]: cast not necessary anymore.

**Review by xchapron-ledger — APPROVED** [review]

File: src/start_new_transaction.c [hunk 2]
```diff
@@ -4,12 +4,19 @@
...
 int start_new_transaction(const command_t *cmd) {
     // prepare size for device_transaction_id + 0x9000
     uint8_t output_buffer[sizeof(G_swap_ctx.device_transaction_id) + 2];
     size_t output_buffer_size = 0;
 
+    // Force a UI reset if we are currently displaying a modal
+    if (G_swap_ctx.state == WAITING_SIGNING) {
+        PRINTF("Dismissing the modal and forcing the main menu\n");
+        ui_idle();
+    }
```
Comment (xchapron-ledger) [inline src/start_new_transaction.c:18]: I'm curious, why is this needed? Won't the next transaction ui screen override the UI if needed?

File: src/ui/validate_transaction_nbgl.c [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+// Delimitor ' ' or '\n'
+
+#define REVIEW_P2 "to"
+
+// Delimitor ' ' or '\n'
+
+// One of:
+#define REVIEW_P3_SWAP "swap "
+#define REVIEW_P3_SELL "sell "
+#define REVIEW_P3_FUND "fund account"
```
Comment (xchapron-ledger) [inline src/ui/validate_transaction_nbgl.c:35]: no space at the end needed here?

**Review by xchapron-ledger — DISMISSED** [review]

File: Makefile [hunk 4]
```diff
@@ -59,42 +63,49 @@ all: default
...
 
-DEFINES       += HAVE_UX_FLOW
-DEFINES       += HAVE_STACK_OVERFLOW_CHECK
-# Enabling debug PRINTF
-
-ifndef DEBUG
-        DEBUG = 0
+ifeq ($(TARGET_NAME),TARGET_STAX)
+    DEFINES += NBGL_QRCODE
+    SDK_SOURCE_PATH += qrcode
```
[the remark this answers is not in the document]
Comment (fbeutin-ledger) [reply Makefile:93]: Removed

File: src/main.c [hunk 5]
```diff
@@ -87,18 +88,37 @@ void app_exit(void) {
...
+    bool was_successful;
+    char appname_last_cycle[BOLOS_APPNAME_MAX_SIZE_B];
+} previous_cycle_data_t;
+
 __attribute__((section(".boot"))) int main(__attribute__((unused)) int arg0) {
     // exit critical section
     __asm volatile("cpsie i");
 
+    // previous_cycle_data needs to be on stack to avoid being erased by our init BSS clean
+    previous_cycle_data_t previous_cycle_data;
```
[the remark this answers is not in the document]
Comment (fbeutin-ledger) [reply src/main.c:102]: Done for BSS + NBGL + simplified this

File: src/start_new_transaction.c (hunk 2 above)
Comment (fbeutin-ledger) [reply src/start_new_transaction.c:18]: Needed if the user has validated the UI (we display a `Processing` modal) but the Live changes it's mind and starts a new transaction flow

File: src/ui/validate_transaction_nbgl.c (hunk 3 above)
Comment (fbeutin-ledger) [reply src/ui/validate_transaction_nbgl.c:35]: No because `dyn_string_1` is empty for FUND, so the next string is the delimitor (` ` or `\n`)
A bit wonky overhaul but it should be done on SDK side, not app side...

**Review by xchapron-ledger — APPROVED** [review]
